cpp/src/arrow/engine/substrait/serde_test.cc

```diff
@@ -0,0 +1,739 @@
+    EXPECT_EQ(*type, *expected_type);
+
+    auto size = ext_set.types().size();
+    ASSERT_OK_AND_ASSIGN(auto serialized, SerializeType(*type, &ext_set));
+    EXPECT_EQ(ext_set.types().size(), size) << "was already added to the set above";
+
+    ASSERT_OK_AND_ASSIGN(auto roundtripped, DeserializeType(*serialized, ext_set));
+    EXPECT_EQ(*roundtripped, *expected_type);
+  }
+}
+
+TEST(Substrait, NamedStruct) {
+  ExtensionSet ext_set;
+
+  auto buf = SubstraitFromJSON("NamedStruct", R"({
+    "struct": {
+      "types": [
+        {"i64": {}},
+        {"list": {"type": {"string":{}} }},
+        {"struct": {
+          "types": [
+            {"fp32": {"nullability": "NULLABILITY_REQUIRED"}},
+            {"string": {}}
+          ]
+        }},
+        {"list": {"type": {"string":{}} }},
+      ]
+    },
+    "names": ["a", "b", "c", "d", "e", "f"]
+  })");
+  ASSERT_OK_AND_ASSIGN(auto schema, DeserializeSchema(*buf, ext_set));
+  Schema expected_schema({
+      field("a", int64()),
+      field("b", list(utf8())),
+      field("c", struct_({
+                     field("d", float32(), /*nullable=*/false),
+                     field("e", utf8()),
+                 })),
+      field("f", list(utf8())),
+  });
+  EXPECT_EQ(*schema, expected_schema);
+
+  ASSERT_OK_AND_ASSIGN(auto serialized, SerializeSchema(*schema, &ext_set));
+  ASSERT_OK_AND_ASSIGN(auto roundtripped, DeserializeSchema(*serialized, ext_set));
+  EXPECT_EQ(*roundtripped, expected_schema);
+
+  // too few names
+  buf = SubstraitFromJSON("NamedStruct", R"({
+    "struct": {"types": [{"i32": {}}, {"i32": {}}, {"i32": {}}]},
+    "names": []
+  })");
+  EXPECT_THAT(DeserializeSchema(*buf, ext_set), Raises(StatusCode::Invalid));
+
+  // too many names
+  buf = SubstraitFromJSON("NamedStruct", R"({
+    "struct": {"types": []},
+    "names": ["a", "b", "c"]
+  })");
+  EXPECT_THAT(DeserializeSchema(*buf, ext_set), Raises(StatusCode::Invalid));
+
+  // no schema metadata allowed
+  EXPECT_THAT(SerializeSchema(Schema({}, key_value_metadata({{"ext", "yes"}})), &ext_set),
+              Raises(StatusCode::Invalid));
+
+  // no schema metadata allowed
+  EXPECT_THAT(
+      SerializeSchema(Schema({field("a", int32(), key_value_metadata({{"ext", "yes"}}))}),
+                      &ext_set),
+      Raises(StatusCode::Invalid));
+}
+
+TEST(Substrait, NoEquivalentArrowType) {
+  auto buf = SubstraitFromJSON("Type", R"({"user_defined_type_reference": 99})");
+  ExtensionSet empty;
+  ASSERT_THAT(
+      DeserializeType(*buf, empty),
+      Raises(StatusCode::Invalid, HasSubstr("did not have a corresponding anchor")));
+}
+
+TEST(Substrait, NoEquivalentSubstraitType) {
+  for (auto type : {
+           date64(),
+           timestamp(TimeUnit::SECOND),
+           timestamp(TimeUnit::NANO),
+           timestamp(TimeUnit::MICRO, "New York"),
+           time32(TimeUnit::SECOND),
+           time32(TimeUnit::MILLI),
+           time64(TimeUnit::NANO),
```

Comment:
Some of these timestamp types we should be able to support.  So I suppose this is another "not round tripped" issue.  I'll add a JIRA.

Reply:
On second thought, what sort of path would these be used?  I think this would only affect users that were producing Substrait plans.  If a plan comes in with a timestamp type and the underlying data is actually seconds then I believe we will do the appropriate cast and handle it.  Since Arrow isn't being used today as a Substrait producer it probably isn't that important we solve this right away.

Reply:
For the timezone-naive types I think it should be as easy as just adding them to the extension YAML and making conversion rules for them, unless I'm missing something here; I had already been wondering why they weren't there when updating the YAML file earlier, but didn't want to feature-creep. For the rest we first need Substrait to support type parameters for user-defined types, as well as string-based type parameters for the timezones. (I suppose if string-based type parameters are a bridge too far we could also define some enum that maps integer parameters to timezones, since it's a finite set at least. Or we could define some set of meta-types representing the possible timezones to use as type argument. But we need type parameters either way.)

Reply:
Ah, ok.  If a user wants to write queries that specifically return or rely on semantics for the courser resolutions they would need Arrow's YAML for it.

Reply:
ARROW-15621